src/hashtable.c

```diff
@@ -633,6 +625,14 @@ static int resize(hashtable *ht, size_t min_capacity, int *malloc_failed) {
         }
     }
 
+    if (ht->type->resizeAllowed) {
+        double fill_factor = (double)min_capacity / ((double)numBuckets(old_exp) * ENTRIES_PER_BUCKET);
+        if (fill_factor * 100 < MAX_FILL_PERCENT_HARD && !ht->type->resizeAllowed(exp > old_exp ? alloc_size : 0, fill_factor)) {
```

Comment:
This effectively ignores the allocation size when shrinking. It means we allow memory usage to temporarily grow above maxmemory. This is risky in a system where maxmemory is set close to the available memory. In the worst case, the kernel's OOM killer will kill Valkey.

Reply:
The `maxmemory` limit is and was never a guarantee that the memory usage of Valkey stays within those bounds, and is also explained in this comment inside `evict.c`:

https://github.com/valkey-io/valkey/blob/4092c9c8e091058e3103007b8dc05ae8ae00c1fc/src/evict.c#L498-L502

Reply:
OK. This is internal comments though, not official documentation. But assuming this is the officially documented behavior, then I guess we can do this change.

Reply:
After talking to @madolson about this PR, I think we can allow it.

Just as we ignore and don't call the callback if we're over MAX_FILL_PERCENT_HARD, we can also ignore it if we're shrinking. I mean this:

```suggestion
        if (fill_factor * 100 < MAX_FILL_PERCENT_HARD &&
            exp > old_exp &&
            !ht->type->resizeAllowed(alloc_size, fill_factor)) {
```

Reply:
I did it this way since `hashtableResizeAllowed` has an additional check `if (!server.dict_resizing) return 0;` (`/* For debug purposes, not allowed to be resized. */`) that I don't want skip.

Reply:
What is it used for? I'm guessing it's only for testing. If it doesn't break any test cases, I think it's fine. Otherwise we can just adjust the tests that are failing.

Reply:
I moved the check up to the `if (ht->type->resizeAllowed && exp > old_exp) {` condition, this test is failing now:
```
[err]: expire scan should skip dictionaries with lot's of empty buckets in tests/unit/expire.tcl
scan didn't handle slot skipping logic.
[1/1 done]: unit/expire (17 seconds)

                   The End

Execution time of different units:
  17 seconds - unit/expire

!!! WARNING The following tests failed:

*** [err]: expire scan should skip dictionaries with lot's of empty buckets in tests/unit/expire.tcl
scan didn't handle slot skipping logic.
```
This test relies on the `dict_resizing` debug flag and in order to fix the test I'd have to move the check for `if (!server.dict_resizing) return 0;` into the `resize` function. That effectively makes the `resizeAllowed` callback function useless. Do we want to move the logic into the `resize` function and get rid of the `resizeAllowed` callback function?

Reply:
There are several (too many) mechanisms to disable resizing. Another is to call `hashtableSetResizePolicy(HASHTABLE_RESIZE_FORBID)`. This is done in `updateDictResizePolicy()` (which is called e.g. after `fork()`).

I suggest that we change it like this:

```diff
 void updateDictResizePolicy(void) {
-    if (server.in_fork_child != CHILD_TYPE_NONE) {
+    if (server.in_fork_child != CHILD_TYPE_NONE || !server.dict_resizing) {
         dictSetResizeEnabled(DICT_RESIZE_FORBID);
         hashtableSetResizePolicy(HASHTABLE_RESIZE_FORBID);
     } else if (hasActiveChildProcess()) {
         dictSetResizeEnabled(DICT_RESIZE_AVOID);
         hashtableSetResizePolicy(HASHTABLE_RESIZE_AVOID);
     } else {
         dictSetResizeEnabled(DICT_RESIZE_ENABLE);
         hashtableSetResizePolicy(HASHTABLE_RESIZE_ALLOW);
     }
 }
```

and in debug.c we call this after setting `dict-resizing`.

```diff
     } else if (!strcasecmp(c->argv[1]->ptr, "dict-resizing") && c->argc == 3) {
         server.dict_resizing = atoi(c->argv[2]->ptr);
+        updateDictResizePolicy();
         addReply(c, shared.ok);
```

With this change, we can delete the check for `server.dict_resizing` in `hashtableResizeAllowed()`.

I think this will be a cleaner implementation. WDYT?

Reply:
That would work but I just realized that we may not be able to get rid of the `hashtableResizeAllowed` callback anyway since I think we're using `hashtable` in non-server stuff as well and calling `overMaxmemoryAfterAlloc()` to check for the `maxmemory` constraint would mean having to pull in `evict.c` and `server.h`.
i do think that the way I initially implemented and PR'd the change was already the most reliable one we could do without introducing a bunch of scary side-effects and unknown gotchas.

Reply:
> we may not be able to get rid of the `hashtableResizeAllowed` callback anyway

That's right. I didn't propose to get rid of `hashtableResizeAllowed` entirely. Only to skip calling it with 0 bytes when we're shrinking is odd and since we're not calling the `resizeAllowed` callback at all when we're over MAX_FILL_PERCENT_HARD, skipping the callback in the same way when shrinking seems more consistent with that.

> and calling `overMaxmemoryAfterAlloc()` to check for the `maxmemory` constraint would mean having to pull in `evict.c` and `server.h`.

No, why would we? The suggested changes in server.c and evict.c only affect server, not other usages such as valkey-benchmark, cli, etc.

> i do think that the way I initially implemented and PR'd the change was already the most reliable one we could do without introducing a bunch of scary side-effects and unknown gotchas.

Are you talking about usages of hashtable outside of this repo?

Reply:
Updated the PR with your proposed changes and included all the implementation changes to make it work. Please let me know if this looks good or if you'd like something changed up. Note that a lot of the changes you see here were necessary to properly merge the `dict-resizing` flag into the `resize_policy` enum and to satisfy the `expire scan should skip dictionaries with lot's of empty buckets` test.